sdk/iot/modelsrepository/src/psuedoParser.ts

```diff
@@ -0,0 +1,94 @@
+// Copyright (c) Microsoft.
+// Licensed under the MIT license.
+
+import { DTDL } from "./DTDL";
+import { logger } from "./logger";
+import { DtmiResolver } from "./dtmiResolver";
+import { RestError } from "@azure/core-rest-pipeline";
+
+export class PseudoParser {
+  private _resolver: DtmiResolver;
+
+  constructor(resolver: DtmiResolver) {
+    this._resolver = resolver;
+  }
+
+  async expand(models: DTDL[], tryFromExpanded: boolean) {
+    let expandedMap: any = {};
+    for (let i=0; i<models.length; i++) {
+      const model: DTDL = models[i];
+      if (model["@id"] !== undefined) {
+        expandedMap[model["@id"]] = model;
+      } else {
+        throw Error(`model ${model} does not contain @id member`);
+      }
+      await this._expand(model, expandedMap, tryFromExpanded);
+    }
+    return expandedMap;
+  }
+
+  private async _expand(model: DTDL, modelMap: any, tryFromExpanded: boolean): Promise<void> {
+    logger.info(`Expanding model: ${model["@id"]}`);
+    let dependencies = this._getModelDependencies(model);
+    let dependenciesToResolve = dependencies.filter((dependency: string) => {
+      return !(dependency in modelMap);
+    });
+    if (dependenciesToResolve.length !== 0) {
+      logger.info(`Outstanding dependencies found: ${dependenciesToResolve}`);
+      let resolvedDependenciesMap: { [s: string]: unknown; };
+      try {
+        resolvedDependenciesMap = await this._resolver.resolve(dependenciesToResolve, tryFromExpanded);
+      } catch (e) {
+        if (e instanceof RestError) {
```

Comment:
instanceof checks have some edge cases. `e.name` should equal `RestError` and avoids these edge cases (most common being if multiple versions of a library gets pulled into your project)

Reply:
Also, can you explain this recovery case? Curious why any RestError and not a subset would cause you to call resolve again with different arguments.

Reply:
`RestError` is the most logical form of Error I could find that would encapsulate the specific error of "I tried to access a model from either a remote or local endpoint and it wasn't there, so either I got a file not found or a 404". Because we're dealing with both file not found or a 404 error, we use a generic wrapper error. If you think there's a better version of naming for this format of Error, I would be open to changing it. In Python I believe it was settled to use "ResourceNotFoundError" but that doesn't exist out of the box in the core or http pipeline libraries.